pulp_rpm/app/serializers/repository.py

```diff
@@ -558,29 +590,83 @@ def validate(self, data):
         Check for cross-domain references (if domain-enabled).
         """
 
-        def check_domain(domain, href, name):
-            # We're doing just a string-check here rather than checking objects
-            # because there can be A LOT of objects, and this is happening in the view-layer
-            # where we have strictly-limited timescales to work with
-            if href and domain not in href:
+        def raise_validation(field, domain, id=""):
+            id = f"\n{id}" if id else ""
+            raise serializers.ValidationError(
+                _("The field {} contains object(s) not in {} domain.{}".format(field, domain, id))
+            )
+
+        def parse_reference(ref) -> tuple[str, str, bool]:
+            """Extract info from prn/href to enable checking domains.
+
+            This is used for:
+            1. In case of HREFS, avoid expensive extract_pk(href) to get pks.
+            2. HREF and PRNs have different information hardcoded available.
+               E.g: RepositoryVerseion HREF has its Repository pk; PRNs have the RepoVer pk.
+
+            Returns a tuple with (pk, class_name, is_prn)
+            """
+            if ref.startswith("prn:"):
+                ref_class, pk = resolve_prn(ref)
+                return (pk, ref_class, True)
+            # content:    ${BASE}/content/rpm/packages/${UUID}/
+            # repository: ${BASE}/repositories/rpm/rpm/${UUID}/
+            # repover:    ${BASE}/repositories/rpm/rpm/${UUID}/versions/0/
+            url = urlparse(ref).path.strip("/").split("/")
+            ref_class = RpmRepository if "/repositories/" in ref else Content
+            is_repover_href = url[-1].isdigit() and url[-2] == "versions"
+            uuid = url[-3] if is_repover_href else url[-1]
+            if len(uuid) < 32:
                 raise serializers.ValidationError(
-                    _("{} must be a part of the {} domain.").format(name, domain)
+                    _("The href path should end with a uuid pk: {}".format(ref))
                 )
+            return (uuid, ref_class, False)
+
+        def check_domain(entry, name, curr_domain):
+            href_or_prn = entry[name]
+            resource_pk, ref_class, is_prn = parse_reference(href_or_prn)
+            try:
+                if ref_class is RepositoryVersion and is_prn:
+                    resource_domain_pk = (
+                        RepositoryVersion.objects.select_related("repository")
+                        .values("repository__pulp_domain")
+                        .get(pk=resource_pk)["repository__pulp_domain"]
+                    )
+                else:
```

Comment:
At the very least this function should have a comment stating that it shouldn't be used on anything but `RpmRepository` and `RepositoryVersion` types.  But it would also be good to throw a `ValueError` if it's not one of those types.